File: golang/146.lru-cache.go
type Node struct {
	Key        int
	Val        int
	Prev, Next *Node
}

type linkedList struct {
	count      int
	head, tail *Node
}

func newLinkedList() *linkedList {
	head, tail := &Node{Key: -1}, &Node{Key: -1}
	head.Next, tail.Prev = tail, head

	return &linkedList{
		count: 0,
		head:  head, // front
		tail:  tail, // back
	}
}

func (l *linkedList) PushFront(node *Node) {
	node.Prev, node.Next = l.head, l.head.Next

	l.head.Next, l.head.Next.Prev = node, node
	l.count++
}

func (l *linkedList) PushBack(node *Node) {
	node.Prev, node.Next = l.tail.Prev, l.tail

	l.tail.Prev, l.tail.Prev.Next = node, node
	l.count++
}

func (l *linkedList) PopFront() *Node {
	tmp := l.head.Next
	l.head.Next.Next.Prev, l.head.Next = l.head, l.head.Next.Next
  l.count--
	return tmp
}

func (l *linkedList) PopBack() *Node {
	tmp := l.tail.Prev
	l.tail.Prev.Prev.Next, l.tail.Prev = l.tail, l.tail.Prev.Prev
  l.count--
	return tmp
}

func (l *linkedList) Remove(node *Node) {
	node.Prev.Next, node.Next.Prev = node.Next, node.Prev
	node.Next, node.Prev = nil, nil
  l.count--
}

func (l *linkedList) PrintFront() {
	for tmp := l.head; tmp != nil; tmp = tmp.Next {
		fmt.Printf("[%d: %d]", tmp.Key, tmp.Val)
	}
	fmt.Println()
}

func (l *linkedList) PrintBack() {
	for tmp := l.tail; tmp != nil; tmp = tmp.Prev {
		fmt.Printf("[%d: %d]", tmp.Key, tmp.Val)
	}
	fmt.Println()
}

type LRUCache struct {
	cap   int
	list  *linkedList
	cache map[int]*Node
}

func Constructor(capacity int) LRUCache {
	return LRUCache{
		cap:   capacity,
		list:  newLinkedList(),
		cache: make(map[int]*Node),
	}
}

func (this *LRUCache) Get(key int) int {

	if node, ok := this.cache[key]; ok {
		this.list.Remove(node)
		this.list.PushFront(node)
		return node.Val
	}

	return -1
}

func (this *LRUCache) Put(key int, value int) {

	if node, ok := this.cache[key]; ok {
		node.Val = value
		this.list.Remove(node)
		this.list.PushFront(node)

		return
	}

	node := &Node{Key: key, Val: value}
	this.cache[key] = node
	this.list.PushFront(node)

	if this.list.count > this.cap {
		toRemove := this.list.PopBack()
		delete(this.cache, toRemove.Key)
	}
}

func (this *LRUCache) Print() {
	fmt.Println(this.cache)
	this.list.PrintFront()
	this.list.PrintBack()
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */

File: python/146.lru-cache.py
class Node:
    slots = ("key", "val", "prev", "next",)

    def __init__(self, key, val):
        self.key, self.val = key, val
        self.prev = self.next = None


class LinkedList:

    def __init__(self):
        self.left, self.right = Node(-1, -1), Node(-1, -1)
        self.left.next, self.right.prev = self.right, self.left

    def remove(self, node: Node) -> None:
        """Remove node from double linked list.
        x - null

                prev     node     nxt
        left ->| n | -> | n | -> | n | -> x
           x <-| p | <- | p | <- | p | <- right
        
        """
        prev, nxt = node.prev, node.next
        prev.next, nxt.prev = nxt, prev

    def insert_right(self, node: Node) -> None:
        """Insert node to LRU before the right pointer.
        x - null

                prev     nxt       
        left ->| n | -> | n | -> x
           x <-| p | <- | p | <- right

                prev     node     nxt
        left ->| n | -> | n | -> | n | -> x
           x <-| p | <- | p | <- | p | <- right
        
        """
        prev, nxt = self.right.prev, self.right
        prev.next, nxt.prev = node, node
        node.prev, node.next = prev, nxt

    def get_left(self) -> Node:
        return self.left.next
        

class LRUCache:

    def __init__(self, capacity: int):
        self.cache: Dict[int, Node] = {}
        self.capacity = capacity
        self.list = LinkedList()

    def get(self, key: int) -> int:
        if key in self.cache:
            node = self.cache[key]
            self.list.remove(node)
            self.list.insert_right(node)
            return node.val
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.list.remove(self.cache[key])

        node = Node(key, value)
        self.cache[key] = node
        self.list.insert_right(node)

        if len(self.cache) > self.capacity:
            lru = self.list.get_left()
            self.list.remove(lru)
            del self.cache[lru.key]


